From c12a075e6adc21968703de33f766f1f60b942a37 Mon Sep 17 00:00:00 2001
From: Andrea Merello <andrea.merello@gmail.com>
Date: Tue, 7 Jul 2020 15:45:51 +0200
Subject: [PATCH 1421/1851] dmaengine: xilinx: frmbuf: add missing controller
 free when removing module

commit b15feae23eb06f562326cff6dcb065c92d4923f7 from
https://github.com/Xilinx/linux-xlnx.git

This fix doesn't just cure a leak; forgetting to release the dma controller
caused Oopses when trying again to get slave channels after rmmod/insmod,
because a dangling controller were accessed in dmaengine core.

NOTE: This has been tested on V2019.1, then rebased. (it look like this
driver didn't change so much indeed).

Signed-off-by: Andrea Merello <andrea.merello@gmail.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/dma/xilinx/xilinx_frmbuf.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/dma/xilinx/xilinx_frmbuf.c b/drivers/dma/xilinx/xilinx_frmbuf.c
index c5ba71991f2b..54dc2c342f90 100644
--- a/drivers/dma/xilinx/xilinx_frmbuf.c
+++ b/drivers/dma/xilinx/xilinx_frmbuf.c
@@ -1684,6 +1684,8 @@ static int xilinx_frmbuf_remove(struct platform_device *pdev)
 {
 	struct xilinx_frmbuf_device *xdev = platform_get_drvdata(pdev);
 
+	of_dma_controller_free(pdev->dev.of_node);
+
 	dma_async_device_unregister(&xdev->common);
 	xilinx_frmbuf_chan_remove(&xdev->chan);
 	clk_disable_unprepare(xdev->ap_clk);
-- 
2.31.1

